File: src/semantic_analysis/Resolve.fs
module Resolve

open Ast.Untyped
open Types

type ResolvedVar = {
    unique_name: string
    from_current_scope: bool
    has_linkage: bool
}

type ResolvedStruct = { unique_tag: string; struct_from_current_scope: bool }

// F#'s Map.map takes (key -> value -> value), unlike OCaml's which takes (value -> value)
let copyIdentifierMap (m: Map<string, ResolvedVar>) =
    Map.map (fun _ entry -> { entry with from_current_scope = false }) m

let copyStructMap m =
    Map.map
        (fun _ entry -> { entry with struct_from_current_scope = false })
        m

(* replace structure tags in type specifiers *)
let rec resolveType struct_map = function
    | Structure tag ->
        (match Map.tryFind tag struct_map with
         | Some entry -> Structure entry.unique_tag
         | None -> failwith "specified undeclared structure type")
    | Pointer referenced_t -> Pointer (resolveType struct_map referenced_t)
    | Array (elem_type, size) ->
        let resolved_elem_type = resolveType struct_map elem_type
        Array (resolved_elem_type, size)
    | FunType (param_types, ret_type) ->
        let resolved_param_types =
            List.map (resolveType struct_map) param_types
        let resolved_ret_type = resolveType struct_map ret_type
        FunType (resolved_param_types, resolved_ret_type)
    | t -> t

let rec resolveExp struct_map id_map = function
    | Exp.Assignment (left, right) ->
        Exp.Assignment
            (resolveExp struct_map id_map left, resolveExp struct_map id_map right)
    | Exp.Var v ->
        (match Map.tryFind v id_map with
         | Some entry -> Exp.Var entry.unique_name
         | None -> failwith (sprintf "Undeclared variable %s" v))
    | Exp.Cast (target_type, e) ->
        let resolved_type = resolveType struct_map target_type
        Exp.Cast (resolved_type, resolveExp struct_map id_map e)
    | Exp.Unary (op, e) -> Exp.Unary (op, resolveExp struct_map id_map e)
    | Exp.Binary (op, e1, e2) ->
        Exp.Binary
            (op, resolveExp struct_map id_map e1, resolveExp struct_map id_map e2)
    | Exp.Conditional (condition, then_result, else_result) ->
        Exp.Conditional
            (resolveExp struct_map id_map condition,
             resolveExp struct_map id_map then_result,
             resolveExp struct_map id_map else_result)
    | Exp.FunCall (f, args) ->
        (match Map.tryFind f id_map with
         | Some entry ->
             Exp.FunCall
                 (entry.unique_name, List.map (resolveExp struct_map id_map) args)
         | None -> failwith "Undeclared function!")
    | Exp.Dereference inner -> Exp.Dereference (resolveExp struct_map id_map inner)
    | Exp.AddrOf inner -> Exp.AddrOf (resolveExp struct_map id_map inner)
    | Exp.Subscript (ptr, index) ->
        Exp.Subscript
            (resolveExp struct_map id_map ptr,
             resolveExp struct_map id_map index)
    | Exp.SizeOf e -> Exp.SizeOf (resolveExp struct_map id_map e)
    | Exp.SizeOfT t -> Exp.SizeOfT (resolveType struct_map t)
    | Exp.Dot (strct, mbr) ->
        Exp.Dot (resolveExp struct_map id_map strct, mbr)
    | Exp.Arrow (strct, mbr) ->
        Exp.Arrow (resolveExp struct_map id_map strct, mbr)
    | (Exp.Constant _ | Exp.String _) as c -> c

let resolveOptionalExp struct_map id_map =
    Option.map (resolveExp struct_map id_map)

let resolveLocalVarHelper counter id_map name storage_class =
    (match Map.tryFind name id_map with
     | Some { from_current_scope = true; has_linkage = has_linkage } ->
         if not (has_linkage && storage_class = Some Ast.StorageClass.Extern) then
             failwith "Duplicate variable declaration"
         else ()
     | _ -> ())
    let counter', entry =
        if storage_class = Some Ast.StorageClass.Extern then
            (counter, { unique_name = name; from_current_scope = true; has_linkage = true })
        else
            let counter', unique_name = UniqueIds.makeNamedTemporary name counter
            (counter', { unique_name = unique_name; from_current_scope = true; has_linkage = false })
    let new_map = Map.add name entry id_map
    (counter', new_map, entry.unique_name)

let rec resolveInitializer struct_map id_map = function
    | Initializer.SingleInit e -> Initializer.SingleInit (resolveExp struct_map id_map e)
    | Initializer.CompoundInit inits ->
        Initializer.CompoundInit (List.map (resolveInitializer struct_map id_map) inits)

let resolveLocalVarDeclaration counter struct_map id_map
        { name = name; varType = var_type; init = init; storageClass = storage_class } =
    let counter', new_id_map, unique_name =
        resolveLocalVarHelper counter id_map name storage_class
    let resolved_type = resolveType struct_map var_type
    let resolved_init =
        Option.map (resolveInitializer struct_map new_id_map) init
    ( counter', new_id_map,
      {
          name = unique_name
          varType = resolved_type
          init = resolved_init
          storageClass = storage_class
      } )

let resolveForInit counter struct_map id_map = function
    | InitExp e -> (counter, id_map, InitExp (resolveOptionalExp struct_map id_map e))
    | InitDecl d ->
        let counter', new_id_map, resolved_decl =
            resolveLocalVarDeclaration counter struct_map id_map d
        (counter', new_id_map, InitDecl resolved_decl)

let rec resolveStatement counter struct_map id_map = function
    | Return e ->
        let resolved_e = Option.map (resolveExp struct_map id_map) e
        (counter, Return resolved_e)
    | Expression e -> (counter, Expression (resolveExp struct_map id_map e))
    | Statement.If (condition, then_clause, else_clause) ->
        let counter', then' = resolveStatement counter struct_map id_map then_clause
        let counter'', else' =
            match else_clause with
            | Some e ->
                let c, e' = resolveStatement counter' struct_map id_map e
                (c, Some e')
            | None -> (counter', None)
        (counter'',
         Statement.If
            (resolveExp struct_map id_map condition,
             then',
             else'))
    | While (condition, body, id) ->
        let counter', body' = resolveStatement counter struct_map id_map body
        (counter',
         While
            (resolveExp struct_map id_map condition,
             body',
             id))
    | DoWhile (body, condition, id) ->
        let counter', body' = resolveStatement counter struct_map id_map body
        (counter',
         DoWhile
            (body',
             resolveExp struct_map id_map condition,
             id))
    | For (init, condition, post, body, id) ->
        let id_map1 = copyIdentifierMap id_map
        let struct_map1 = copyStructMap struct_map
        let counter', id_map2, resolved_init = resolveForInit counter struct_map1 id_map1 init
        let counter'', body' = resolveStatement counter' struct_map1 id_map2 body
        (counter'',
         For
            (resolved_init,
             resolveOptionalExp struct_map1 id_map2 condition,
             resolveOptionalExp struct_map1 id_map2 post,
             body',
             id))
    | Compound block ->
        let new_variable_map = copyIdentifierMap id_map
        let new_struct_map = copyStructMap struct_map
        let counter', block' = resolveBlock counter new_struct_map new_variable_map block
        (counter', Compound block')
    | (Null | Break _ | Continue _) as s -> (counter, s)

and resolveBlockItem counter (struct_map, id_map) = function
    | Stmt s ->
        let counter', resolved_s = resolveStatement counter struct_map id_map s
        (counter', (struct_map, id_map), Stmt resolved_s)
    | Decl d ->
        let counter', new_maps, resolved_d =
            resolveLocalDeclaration counter struct_map id_map d
        (counter', new_maps, Decl resolved_d)

and resolveBlock counter struct_map id_map (Block items) =
    let counter', _final_maps, resolved_items =
        List.fold (fun (c, maps, acc) item ->
            let c', maps', item' = resolveBlockItem c maps item
            (c', maps', acc @ [item'])) (counter, (struct_map, id_map), []) items
    (counter', Block resolved_items)

and resolveLocalDeclaration counter struct_map id_map = function
    | VarDecl vd ->
        let counter', new_id_map, resolved_vd =
            resolveLocalVarDeclaration counter struct_map id_map vd
        (counter', (struct_map, new_id_map), VarDecl resolved_vd)
    | FunDecl { body = Some _ } ->
        failwith "nested function definitions are not allowed"
    | FunDecl { storageClass = Some Ast.StorageClass.Static } ->
        failwith "static keyword not allowed on local function declarations"
    | FunDecl fd ->
        let counter', new_id_map, resolved_fd =
            resolveFunctionDeclaration counter struct_map id_map fd
        (counter', (struct_map, new_id_map), FunDecl resolved_fd)
    | StructDecl sd ->
        let counter', new_struct_map, resolved_sd =
            resolveStructureDeclaration counter struct_map sd
        (counter', (new_struct_map, id_map), StructDecl resolved_sd)

and resolveParams counter id_map param_names =
    List.fold (fun (c, m, acc) param_name ->
        let c', m', unique = resolveLocalVarHelper c m param_name None
        (c', m', acc @ [unique])) (counter, id_map, []) param_names

and resolveFunctionDeclaration counter struct_map id_map fn =
    match Map.tryFind fn.name id_map with
    | Some { from_current_scope = true; has_linkage = false } ->
        failwith "Duplicate declaration"
    | _ ->
        let resolved_type = resolveType struct_map fn.funType
        let new_entry =
            { unique_name = fn.name; from_current_scope = true; has_linkage = true }
        let new_id_map = Map.add fn.name new_entry id_map
        let inner_id_map = copyIdentifierMap new_id_map
        let counter', inner_id_map1, resolved_params =
            resolveParams counter inner_id_map fn.``params``
        let inner_struct_map = copyStructMap struct_map
        let counter'', resolved_body =
            match fn.body with
            | Some body ->
                let c, b = resolveBlock counter' inner_struct_map inner_id_map1 body
                (c, Some b)
            | None -> (counter', None)
        ( counter'', new_id_map,
          {
              fn with
                  funType = resolved_type
                  ``params`` = resolved_params
                  body = resolved_body
          } )

and resolveStructureDeclaration counter struct_map { tag = tag; members = members } =
    let prev_entry = Map.tryFind tag struct_map
    let counter', new_map, resolved_tag =
        match prev_entry with
        | Some { unique_tag = unique_tag; struct_from_current_scope = true } ->
            (counter, struct_map, unique_tag)
        | _ ->
            let counter', unique_tag = UniqueIds.makeNamedTemporary tag counter
            let entry = { unique_tag = unique_tag; struct_from_current_scope = true }
            (counter', Map.add tag entry struct_map, unique_tag)
    let resolveMember m =
        { m with memberType = resolveType new_map m.memberType }
    let resolved_members = List.map resolveMember members
    (counter', new_map, { tag = resolved_tag; members = resolved_members })

let resolveFileScopeVariableDeclaration struct_map id_map
        ({ name = name; varType = var_type } as vd: VariableDeclaration) =
    let resolved_vd = { vd with varType = resolveType struct_map var_type }
    let new_map =
        Map.add name
            { unique_name = name; from_current_scope = true; has_linkage = true }
            id_map
    (new_map, resolved_vd)

let resolveGlobalDeclaration counter (struct_map, id_map) = function
    | FunDecl fd ->
        let counter', id_map1, fd = resolveFunctionDeclaration counter struct_map id_map fd
        (counter', (struct_map, id_map1), FunDecl fd)
    | VarDecl vd ->
        let id_map1, resolved_vd =
            resolveFileScopeVariableDeclaration struct_map id_map vd
        (counter, (struct_map, id_map1), VarDecl resolved_vd)
    | StructDecl sd ->
        let counter', struct_map1, resolved_sd =
            resolveStructureDeclaration counter struct_map sd
        (counter', (struct_map1, id_map), StructDecl resolved_sd)

let resolve counter (Program decls) =
    let counter', _, resolved_decls =
        List.fold (fun (c, maps, acc) d ->
            let c', maps', d' = resolveGlobalDeclaration c maps d
            (c', maps', acc @ [d'])) (counter, (Map.empty, Map.empty), []) decls
    (counter', Program resolved_decls)

File: src/semantic_analysis/Resolve.rs
//! Resolve source identifiers and structure tags to unique names.
#![allow(non_snake_case)]

use std::collections::BTreeMap;
use super::{Ast::*, Types::Type, UniqueIds};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ResolvedVar { pub unique_name: String, pub from_current_scope: bool, pub has_linkage: bool }
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ResolvedStruct { pub unique_tag: String, pub struct_from_current_scope: bool }
type IdMap = BTreeMap<String, ResolvedVar>;
type StructMap = BTreeMap<String, ResolvedStruct>;

fn copied_ids(map: &IdMap) -> IdMap { map.iter().map(|(k,v)| (k.clone(), ResolvedVar { from_current_scope:false, ..v.clone() })).collect() }
fn copied_structs(map: &StructMap) -> StructMap { map.iter().map(|(k,v)| (k.clone(), ResolvedStruct { struct_from_current_scope:false, ..v.clone() })).collect() }

pub fn resolveType(structs: &StructMap, ty: Type) -> Type { match ty {
    Type::Struct(tag) => Type::Struct(structs.get(&tag).unwrap_or_else(|| panic!("specified undeclared structure type")).unique_tag.clone()),
    Type::Pointer(t) => Type::Pointer(Box::new(resolveType(structs, *t))),
    Type::Array(t,n) => Type::Array(Box::new(resolveType(structs,*t)),n),
    Type::Function(ps,r) => Type::Function(ps.into_iter().map(|t|resolveType(structs,t)).collect(), Box::new(resolveType(structs,*r))),
    t => t,
} }
fn exp(structs:&StructMap, ids:&IdMap, e:Exp)->Exp { match e {
    Exp::Var(v)=>Exp::Var(ids.get(&v).unwrap_or_else(||panic!("Undeclared variable {v}")).unique_name.clone()),
    Exp::Assignment(a,b)=>Exp::Assignment(Box::new(exp(structs,ids,*a)),Box::new(exp(structs,ids,*b))),
    Exp::Cast(t,e)=>Exp::Cast(resolveType(structs,t),Box::new(exp(structs,ids,*e))),
    Exp::Unary(o,e)=>Exp::Unary(o,Box::new(exp(structs,ids,*e))),
    Exp::Binary(o,a,b)=>Exp::Binary(o,Box::new(exp(structs,ids,*a)),Box::new(exp(structs,ids,*b))),
    Exp::Conditional(a,b,c)=>Exp::Conditional(Box::new(exp(structs,ids,*a)),Box::new(exp(structs,ids,*b)),Box::new(exp(structs,ids,*c))),
    Exp::FunCall(f,args)=>Exp::FunCall(ids.get(&f).unwrap_or_else(||panic!("Undeclared function!")).unique_name.clone(),args.into_iter().map(|e|exp(structs,ids,e)).collect()),
    Exp::Dereference(e)=>Exp::Dereference(Box::new(exp(structs,ids,*e))), Exp::AddrOf(e)=>Exp::AddrOf(Box::new(exp(structs,ids,*e))),
    Exp::Subscript(a,b)=>Exp::Subscript(Box::new(exp(structs,ids,*a)),Box::new(exp(structs,ids,*b))),
    Exp::SizeOf(e)=>Exp::SizeOf(Box::new(exp(structs,ids,*e))), Exp::SizeOfT(t)=>Exp::SizeOfT(resolveType(structs,t)),
    Exp::Dot(e,m)=>Exp::Dot(Box::new(exp(structs,ids,*e)),m), Exp::Arrow(e,m)=>Exp::Arrow(Box::new(exp(structs,ids,*e)),m),
    e => e,
} }
fn initializer(s:&StructMap,i:&IdMap,x:Initializer)->Initializer { match x { Initializer::SingleInit(e)=>Initializer::SingleInit(exp(s,i,e)), Initializer::CompoundInit(xs)=>Initializer::CompoundInit(xs.into_iter().map(|x|initializer(s,i,x)).collect()) } }
fn local_var(counter:usize,s:&StructMap,mut ids:IdMap,mut d:VariableDeclaration)->(usize,IdMap,VariableDeclaration) {
    if let Some(old)=ids.get(&d.name) { if old.from_current_scope && !(old.has_linkage && d.storageClass==Some(StorageClass::Extern)) { panic!("Duplicate variable declaration") } }
    let (counter,unique,linkage)=if d.storageClass==Some(StorageClass::Extern){(counter,d.name.clone(),true)}else{let(c,n)=UniqueIds::make_named_temporary(&d.name,counter);(c,n,false)};
    ids.insert(d.name.clone(),ResolvedVar{unique_name:unique.clone(),from_current_scope:true,has_linkage:linkage});
    d.name=unique; d.varType=resolveType(s,d.varType); d.init=d.init.map(|x|initializer(s,&ids,x)); (counter,ids,d)
}
fn function(counter:usize,s:&StructMap,mut ids:IdMap,mut f:FunctionDeclaration)->(usize,IdMap,FunctionDeclaration) {
    if matches!(ids.get(&f.name),Some(ResolvedVar{from_current_scope:true,has_linkage:false,..})){panic!("Duplicate declaration")}
    f.funType=resolveType(s,f.funType); ids.insert(f.name.clone(),ResolvedVar{unique_name:f.name.clone(),from_current_scope:true,has_linkage:true});
    let mut inner=copied_ids(&ids); let mut counter=counter; let mut params=Vec::new();
    for p in f.params { let dummy=VariableDeclaration{name:p,varType:Type::Int,init:None,storageClass:None}; let(c,m,d)=local_var(counter,s,inner,dummy);counter=c;inner=m;params.push(d.name); }
    f.params=params; f.body=f.body.map(|b|{let(c,b)=block(counter,&copied_structs(s),inner.clone(),b);counter=c;b}); (counter,ids,f)
}
fn structure(counter:usize,mut s:StructMap,mut d:StructDeclaration)->(usize,StructMap,StructDeclaration){
    let (counter,tag)=match s.get(&d.tag){Some(x) if x.struct_from_current_scope=>(counter,x.unique_tag.clone()),_=>{let(c,t)=UniqueIds::make_named_temporary(&d.tag,counter);s.insert(d.tag.clone(),ResolvedStruct{unique_tag:t.clone(),struct_from_current_scope:true});(c,t)}};
    d.tag=tag; for m in &mut d.members {m.memberType=resolveType(&s,m.memberType.clone())} (counter,s,d)
}
fn statement(counter:usize,s:&StructMap,ids:&IdMap,x:Statement)->(usize,Statement){match x{
    Statement::Return(e)=>(counter,Statement::Return(e.map(|e|exp(s,ids,e)))), Statement::Expression(e)=>(counter,Statement::Expression(exp(s,ids,e))),
    Statement::If(e,a,b)=>{let(c,a)=statement(counter,s,ids,*a);let(c,b)=match b{Some(b)=>{let(c,b)=statement(c,s,ids,*b);(c,Some(Box::new(b)))},None=>(c,None)};(c,Statement::If(exp(s,ids,e),Box::new(a),b))},
    Statement::While(e,b,id)=>{let(c,b)=statement(counter,s,ids,*b);(c,Statement::While(exp(s,ids,e),Box::new(b),id))},
    Statement::DoWhile(b,e,id)=>{let(c,b)=statement(counter,s,ids,*b);(c,Statement::DoWhile(Box::new(b),exp(s,ids,e),id))},
    Statement::For(init,cond,post,b,id)=>{let s1=copied_structs(s);let ids1=copied_ids(ids);let(c,ids2,init)=match init{ForInit::InitExp(e)=>(counter,ids1,ForInit::InitExp(e.map(|e|exp(&s1,ids,e)))),ForInit::InitDecl(d)=>{let(c,m,d)=local_var(counter,&s1,ids1,d);(c,m,ForInit::InitDecl(d))}};let(c,b)=statement(c,&s1,&ids2,*b);(c,Statement::For(init,cond.map(|e|exp(&s1,&ids2,e)),post.map(|e|exp(&s1,&ids2,e)),Box::new(b),id))},
    Statement::Compound(b)=>{let(c,b)=block(counter,&copied_structs(s),copied_ids(ids),b);(c,Statement::Compound(b))}, x=>(counter,x)}}
fn block(counter:usize,s:&StructMap,ids:IdMap,b:Block)->(usize,Block){let(mut c,mut sm,mut im)=(counter,s.clone(),ids);let mut out=Vec::new();for x in b.0{match x{BlockItem::Stmt(x)=>{let(n,x)=statement(c,&sm,&im,x);c=n;out.push(BlockItem::Stmt(x))},BlockItem::Decl(d)=>{let(n,ns,ni,d)=local_decl(c,sm,im,d);c=n;sm=ns;im=ni;out.push(BlockItem::Decl(d))}}}(c,Block(out))}
fn local_decl(c:usize,s:StructMap,i:IdMap,d:Declaration)->(usize,StructMap,IdMap,Declaration){match d{
    Declaration::VarDecl(d)=>{let(c,i,d)=local_var(c,&s,i,d);(c,s,i,Declaration::VarDecl(d))},
    Declaration::FunDecl(f) if f.body.is_some()=>panic!("nested function definitions are not allowed"),
    Declaration::FunDecl(f) if f.storageClass==Some(StorageClass::Static)=>panic!("static keyword not allowed on local function declarations"),
    Declaration::FunDecl(f)=>{let(c,i,f)=function(c,&s,i,f);(c,s,i,Declaration::FunDecl(f))},
    Declaration::StructDecl(d)=>{let(c,s,d)=structure(c,s,d);(c,s,i,Declaration::StructDecl(d))}}}
/// Port of `resolve`, threading the unique-id counter explicitly.
pub fn resolve(counter:usize,p:UntypedProgram)->(usize,UntypedProgram){let(mut c,mut s,mut i)=(counter,StructMap::new(),IdMap::new());let mut out=Vec::new();for d in p.0{match d{
    Declaration::FunDecl(f)=>{let(n,ni,f)=function(c,&s,i,f);c=n;i=ni;out.push(Declaration::FunDecl(f))},
    Declaration::VarDecl(mut d)=>{d.varType=resolveType(&s,d.varType);i.insert(d.name.clone(),ResolvedVar{unique_name:d.name.clone(),from_current_scope:true,has_linkage:true});out.push(Declaration::VarDecl(d))},
    Declaration::StructDecl(d)=>{let(n,ns,d)=structure(c,s,d);c=n;s=ns;out.push(Declaration::StructDecl(d))}}}(c,UntypedProgram(out))}
